packages/react-positioning/src/usePopper.ts

```diff
@@ -298,86 +110,103 @@ export function usePopper(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   arrowRef: React.MutableRefObject<any>;
 } {
-  const { enabled = true } = options;
   const isFirstMount = useFirstMount();
+  const { targetDocument } = useFluent();
+  const { enabled = true } = options;
+  const resolvePopperOptions = usePopperOptions(options);
 
-  const popperOriginalPositionRef = React.useRef<string>('absolute');
-  const resolvePopperOptions = usePopperOptions(options, popperOriginalPositionRef);
+  const forceUpdate = useEventCallback(() => {
+    const target = overrideTargetRef.current ?? targetRef.current;
+    if (!canUseDOM || !enabled || !target || !containerRef.current) {
+      return;
+    }
 
-  const popperInstanceRef = React.useRef<PopperInstance | null>(null);
+    const { placement, middleware, strategy } = resolvePopperOptions(target, containerRef.current, arrowRef.current);
 
-  const handlePopperUpdate = useEventCallback(() => {
-    popperInstanceRef.current?.destroy();
-    popperInstanceRef.current = null;
+    Object.assign(containerRef.current.style, { position: strategy });
+    computePosition(target, containerRef.current, { placement, middleware, strategy }).then(
+      ({ x, y, middlewareData, placement: computedPlacement }) => {
+        if (!containerRef.current) {
+          return;
+        }
 
-    const target = overrideTargetRef.current ?? targetRef.current;
+        containerRef.current.setAttribute('data-popper-placement', computedPlacement);
+        Object.assign(containerRef.current.style, {
+          // Layer acceleration can disable subpixel rendering which causes slightly
+          // blurry text on low PPI displays, so we want to use 2D transforms
+          // instead
+          transform:
+            (targetDocument?.defaultView?.devicePixelRatio || 1) <= 1
+              ? `translate(${x}px, ${y}px)`
+              : `translate3d(${x}px, ${y}px, 0)`,
+          position: strategy,
+        });
```

Comment:
All assignments there are hard to read and follow, can we add functions around them?

For example:
```ts
applyStyles()
applyDataAttributes()
applyArrowStyles()
```